### ISSUE TITLE
Entity Adapter

### ISSUE THREAD

**Issue by amcdnl** [ngxs/store#56]
A while back I prototyped something like John Papas NGRX-DATA, https://gist.github.com/amcdnl/cafd37bc9a99e2cd7653ebc023e06ffe#file-readme-md

Something we might want to think about one day.

**Comment by eugene-s** [comment]
Yes, it would be nice. Entity adapter gives more comfortable things to manipulate with storage.

**Comment by wesleygrimes** [comment]
I am switching over from NGRX where I heavily used the NGRX Entity modules. Would love to see something like this implemented for NGXS.

**Comment by filipjnc** [comment]
How far in the future is the entity adapter planned for? This is the only reason why I’m hesitating to switch. The feeling sucks because I can’t wait to switch.

**Comment by abalad** [comment]
Is there a provision for such an appeal? I'd love to see this in NGXS

**Comment by dfulgham** [comment]
When patching your state instead of just dropping the array of values (data) use:

```
store.patchState({ entities: Object.assign(
            {},
            ...data.map(item => ({ [item["id"]]: item }))
          )})
```
In ngrx-data they also have an array of id's, so you could add that as well.

```
store.patchState({ 
    entities: Object.assign(
            {},
            ...data.map(item => ({ [item["id"]]: item })),
    ids: ...data.map(item => item["id"] })
})
```

**Comment by abalad** [comment]
I have created a LIB to be used as follows. I'm using it on mine.
projects.

https://github.com/ngxs/store/issues/541#issuecomment-412874263

**Comment by markwhitfeld** [comment]
@abalad I would love to see what this lib looks like!

**Comment by abalad** [comment]
is only a class with static methods for now, but I want to do better integration within NGXS:

**Example:** https://gist.github.com/abalad/d9c82af8bcdce904bb173a7e1fe3edaa

**Lib:** 

- https://gist.github.com/abalad/c161312b4bb7a4cb441ac6c3f24093ef
- https://gist.github.com/abalad/b400021eb67cd77179c266341a09e561

**Comment by juracy** [comment]
A proposal for entity through decorator: #580 
I would like to discuss the idea!

**Comment by splincode** [comment]
https://github.com/ngxs-labs/entity-state/issues/15
